restock-shop.lic

```diff
@@ -12,98 +12,95 @@ class ShopRestock
+    @bag = @settings.crafting_container
+    # TODO: Validate users item list against recipe data. Ensure that there is only one match per user specified recipe.
+    # TODO: Validate users town. Allow only towns with plazas for simplicity
+
+    # @item_list = [
+    #   ['backpack', 100_000, 5, 'on table', 3, 'backpack', 'heavy linen'],
+    #   ['bag', 100_000, 6, 'on table', 3, 'duffel bag', 'heavy linen'],
+    #   ['sealed', 220_000, 8, 'in cabinet', 'gryphon'],
+    #   ['sword', 140_000, 9, 'on rack', 'bastard sword', 'steel'],
+    #   ['mace', 140_000, 12, 'on rack', 'bar mace', 'steel'],
+    #   ['hammer', 120_000, 10, 'on rack', 'throwing hammer', 'steel'],
+    #   ['greataxe', 120_000, 12, 'on rack', 'greataxe', 'steel'],
+    #   ['carryall', 200_000, 5, 'on table', 3, 'carryall', 'heavy silk'],
+    #   ['haversack', 200_000, 6, 'on table', 3, 'haversack', 'heavy silk']
+    # ]
+    go_to_shop
     check_surfaces
 
-    return if Room.current.id == @inner_id
-    walk_to(@shop_id)
-    fput('open entrance')
-    pause 0.5
-    fput('go entrance')
+    go_to_shop
   end
 
   def check_surfaces
     missing_items = []
-    @item_list.map { |x| x[3] }.uniq.each do |surface|
+    @item_list.map { |data| data['surface'] }.uniq.each do |surface|
       raw_list = bput("look #{surface}", "There's nothing", 'you see [^\.]*')
       raw_list = if raw_list == "There's nothing"
                    []
                  else
-                   list_to_nouns(raw_list)
+                   list_to_array(raw_list)
                  end
-      missing_items += @item_list.select { |data| surface == data[3] }.reject { |data| raw_list.include?(data.first) }
+      missing_items += @item_list.select { |item| surface == item['surface'] }
+                                 .reject { |item| raw_list.find { |raw_item| raw_item.match(item['full_name']) } }
     end
+
+    return if missing_items.empty?
+
+    respond 'Restocking'
+    respond '------------'
+    missing_items.each { |item| respond "- #{item['full_name']}" }
+
+    all_recipes = get_data('recipes').crafting_recipes
+    missing_recipes_re = Regexp.union(missing_items.map { |item| item['recipe'] })
+    @recipes = all_recipes.select { |recipe| missing_recipes_re.match(recipe['name']) }
+
+    ensure_copper_on_hand(40_000, @settings)
     restock_surfaces(missing_items)
   end
 
   def restock_surfaces(missing_items)
-    echo missing_items
-    restock_steel(missing_items.select { |x| x.last == 'steel' })
-    restock_burlap(missing_items.select { |x| x.last == 'heavy burlap' })
-    restock_silk(missing_items.select { |x| x.last == 'heavy silk' })
-    restock_linen(missing_items.select { |x| x.last == 'heavy linen' })
-    restock_gryphon(missing_items.select { |x| x.last == 'gryphon' })
+    restock_steel(missing_items.select { |item| item['material'] == 'steel' })
+    restock_burlap(missing_items.select { |item| item['material'] == 'heavy burlap' })
```

Comment:
💡 How much overlap is there in the cloth restocking functions? I.e. could we do something like `restock_cloth(missing_items, 'burlap', 'heavy burlap')`?

Reply:
I do think they can be combined into some general procedure, I'd like to save that for a future update - I'll have more outfitting ranks to work with (Forging and burlap armor making are currently the only two bits that have been heavily tested so far)

Reply:
👍